File: typescript/src/ws/connection.ts
/**
 * WsConnection — wraps an IWebSocket, implements ITransportConnection.
 * Handles ping/pong heartbeat on the server side.
 */

import type { IWebSocket } from "../core.js";
import { DEFAULT_PING_INTERVAL, DEFAULT_PONG_TIMEOUT } from "../core.js";
import type { ITransportConnection } from "../transport.js";

const WS_OPEN = 1;
const WS_CLOSING = 2;

export class WsConnection implements ITransportConnection {
	onMessage: ((raw: string) => void) | null = null;
	onClose: (() => void) | null = null;

	private _closed = false;
	private _pingTimer: ReturnType<typeof setInterval> | null = null;
	private _pongTimer: ReturnType<typeof setTimeout> | null = null;

	constructor(
		public readonly ws: IWebSocket,
		private readonly pingInterval: number = DEFAULT_PING_INTERVAL,
	) {}

	get isOpen(): boolean {
		return !this._closed && this.ws.readyState === WS_OPEN;
	}

	send(raw: string): void {
		if (this.ws.readyState === WS_OPEN) {
			this.ws.send(raw);
		}
	}

	close(): void {
		this._closed = true;
		this._stopPing();
		if (this.ws.readyState === WS_OPEN || this.ws.readyState === WS_CLOSING) {
			this.ws.close();
		}
	}

	/**
	 * Start the message loop and heartbeat.
	 * Returns a promise that resolves when the connection closes.
	 * Used on the server side to await connection lifetime.
	 */
	serve(): Promise<void> {
		return new Promise<void>((resolve) => {
			this._startPing();

			this.ws.onmessage = (ev: { data: unknown }) => {
				this.onMessage?.(String(ev.data));
			};

			this.ws.onclose = () => {
				this._closed = true;
				this._stopPing();
				this.onClose?.();
				resolve();
			};

			this.ws.onerror = () => {};
		});
	}

	private _startPing(): void {
		this._stopPing();
		this._pingTimer = setInterval(() => {
			if (this.ws.readyState === WS_OPEN) {
				this.ws.send(JSON.stringify({ jsonrpc: "2.0", method: "ping" }));
				this._armPongTimeout();
			}
		}, this.pingInterval);
	}

	private _armPongTimeout(): void {
		if (this._pongTimer) return;
		this._pongTimer = setTimeout(() => {
			this._pongTimer = null;
			this.close();
		}, DEFAULT_PONG_TIMEOUT);
	}

	/** Clear pong timeout (called when pong is received via the router). */
	refreshPong(): void {
		if (this._pongTimer) {
			clearTimeout(this._pongTimer);
			this._pongTimer = null;
		}
	}

	private _stopPing(): void {
		if (this._pingTimer) {
			clearInterval(this._pingTimer);
			this._pingTimer = null;
		}
		if (this._pongTimer) {
			clearTimeout(this._pongTimer);
			this._pongTimer = null;
		}
	}
}

File: python/viberpc/ws/connection.py
"""WsConnection — wraps a websockets connection, provides ITransportConnection."""

from __future__ import annotations

import asyncio
import json
import logging
from typing import Any, Callable

import websockets
from websockets.asyncio.server import ServerConnection
from websockets.asyncio.client import ClientConnection

from ..core import DEFAULT_PING_INTERVAL, PONG_TIMEOUT, make_notification

logger = logging.getLogger("viberpc")

WebSocketConn = ServerConnection | ClientConnection


class WsConnection:
    """Wraps a single WebSocket and provides transport-level I/O."""

    def __init__(
        self,
        ws: WebSocketConn,
        *,
        ping_interval: float = DEFAULT_PING_INTERVAL,
    ) -> None:
        self.ws = ws
        self.ping_interval = ping_interval
        self._closed = False
        self._ping_task: asyncio.Task[None] | None = None
        self._pong_timer: asyncio.TimerHandle | None = None

        self.on_message: Callable[[str], Any] | None = None
        self.on_close: Callable[[], Any] | None = None

    @property
    def is_open(self) -> bool:
        return not self._closed and self.ws.state.name == "OPEN"

    async def send(self, raw: str) -> None:
        """Send raw string."""
        if self._closed:
            return
        await self.ws.send(raw)

    async def close(self) -> None:
        self._closed = True
        self._cancel_ping()
        self._cancel_pong_timer()
        try:
            await self.ws.close()
        except Exception:
            pass

    async def serve(self) -> None:
        """Start listening + heartbeat. Blocks until connection closes."""
        self._ping_task = asyncio.create_task(self._ping_loop())
        try:
            async for raw in self.ws:
                if self.on_message:
                    result = self.on_message(raw)
                    if asyncio.iscoroutine(result):
                        await result
        except websockets.ConnectionClosed:
            pass
        finally:
            self._closed = True
            self._cancel_ping()
            self._cancel_pong_timer()
            if self.on_close:
                result = self.on_close()
                if asyncio.iscoroutine(result):
                    await result

    async def _ping_loop(self) -> None:
        try:
            while not self._closed:
                await asyncio.sleep(self.ping_interval)
                if self._closed:
                    break
                try:
                    await self.ws.send(json.dumps(make_notification("ping")))
                    self._arm_pong_timeout()
                except Exception:
                    break
        except asyncio.CancelledError:
            pass

    def _arm_pong_timeout(self) -> None:
        if self._pong_timer:
            return
        loop = asyncio.get_running_loop()
        self._pong_timer = loop.call_later(PONG_TIMEOUT, self._pong_expired)

    def _pong_expired(self) -> None:
        self._pong_timer = None
        asyncio.ensure_future(self.close())

    def refresh_pong(self) -> None:
        """Clear pong timeout (called when pong is received via the router)."""
        self._cancel_pong_timer()

    def _cancel_pong_timer(self) -> None:
        if self._pong_timer:
            self._pong_timer.cancel()
            self._pong_timer = None

    def _cancel_ping(self) -> None:
        if self._ping_task and not self._ping_task.done():
            self._ping_task.cancel()
